instrumentation/kafka-clients-metrics-3.0.0/src/main/java/com/nr/instrumentation/kafka/NodeTopicRegistry.java

```diff
@@ -0,0 +1,92 @@
+/*
+ *
+ *  * Copyright 2025 New Relic Corporation. All rights reserved.
+ *  * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+package com.nr.instrumentation.kafka;
+
+import org.apache.kafka.common.Node;
+import org.apache.kafka.common.metrics.KafkaMetric;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.nr.instrumentation.kafka.MetricsConstants.NODE_PREFIX;
+
+/**
+ * This class is used to track all the metric names that are related to a specific node:
+ *
+ * - MessageBroker/Kafka/Nodes/host:port
+ * - MessageBroker/Kafka/Nodes/host:port/Consume/topicName
+ * - MessageBroker/Kafka/Nodes/host:port/Produce/topicName
+ *
+ * At initialization time we only have the node and the mode (is this a metrics reporter
+ * for a Kafka consumer or for a Kafka producer?).
+ *
+ * Then, as topics are discovered through the metricChange method, the topic metric names are
+ * generated. This is the best way we have to get track of the topics since they're not
+ * available when the KafkaConsumer/KafkaProducer is initialized.
+ *
+ * For KafkaConsumer, the SubscriptionState doesn't contain the topics and partitions
+ * at initialization time because it takes time for the rebalance to happen.
+ *
+ * For KafkaProducer, topics are dynamic since a producer could send records to any
+ * topic and the concept of subscription doesn't exist there.
+ *
+ * Alternatively we could get the topics from the records in KafkaProducer.doSend or
+ * KafkaConsumer.poll, and call NewRelicMetricsReporter.addTopicToNodeMetrics from there.
+ * This approach would have a small impact in performance, and getting the topics from the
+ * KafkaMetrics is a good enough solution.
+ */
+public class NodeTopicRegistry {
+    private final Set<String> recordedTopics = ConcurrentHashMap.newKeySet();
+    // contains the registered metric or event names, according to config
+    private final Set<String> convertedNames = ConcurrentHashMap.newKeySet();
+    private final Set<String> nodes = new HashSet<>();
```

Comment:
This field is not altered after the constructor, so not bothering to add concurrency control to it.